yls/snippets.py

```diff
@@ -0,0 +1,122 @@
+    config = config or {}
+
+    res = []
+    if config.get("meta", False):
+        item = lsp_types.CompletionItem(
+            label="meta",
+            detail="Generate a meta section (YARA)",
+            kind=lsp_types.CompletionItemKind.Snippet,
+            insert_text=str(_generate_meta_snippet(SnippetString(), config)),
+            insert_text_format=lsp_types.InsertTextFormat.Snippet,
+            documentation=utils.markdown_content('meta:\n\tKEY = "VALUE"'),
+            sort_text="a",
+        )
+        res.append(item)
+
+    if config.get("condition", False):
+        item = lsp_types.CompletionItem(
+            label="condition",
+            detail="Generate a condition section (YARA)",
+            kind=lsp_types.CompletionItemKind.Snippet,
+            insert_text=str(_generate_condition_snippet(SnippetString())),
+            insert_text_format=lsp_types.InsertTextFormat.Snippet,
+            documentation=utils.markdown_content("condition:\n\tCONDITIONS"),
+            sort_text="a",
+        )
+        res.append(item)
+
+    if config.get("strings", False):
+        item = lsp_types.CompletionItem(
+            label="strings",
+            detail="Generate a strings skeleton (YARA)",
+            kind=lsp_types.CompletionItemKind.Snippet,
+            insert_text=str(_generate_string_snippet(SnippetString())),
+            insert_text_format=lsp_types.InsertTextFormat.Snippet,
+            documentation=utils.markdown_content('strings:\n\t$NAME = "STRING"'),
+            sort_text="a",
+        )
+        res.append(item)
+
+    if config.get("rule", False):
+        item = lsp_types.CompletionItem(
+            label="rule",
+            detail="Generate a rule skeleton (YARA)",
+            kind=lsp_types.CompletionItemKind.Snippet,
+            insert_text=str(_generate_rule_snippet(SnippetString(), config)),
+            insert_text_format=lsp_types.InsertTextFormat.Snippet,
+            documentation=utils.markdown_content("rule NAME {"),
+            sort_text="a",
+        )
+        res.append(item)
+
+    return res
+
+
+def _generate_rule_snippet(snippet: SnippetString, config: dict) -> SnippetString:
+    snippet.append_text("rule ")
+    snippet.append_placeholder("my_rule")
+    snippet.append_text(" {\n")
+    snippet = _generate_meta_snippet(snippet, config)
+    snippet = _generate_string_snippet(snippet)
+    snippet.append_text("\n")
+    snippet = _generate_condition_snippet(snippet)
+    snippet.append_text("\n}\n")
+    return snippet
+
+
+def _generate_condition_snippet(snippet: SnippetString) -> SnippetString:
+    snippet.append_text("\tcondition:\n\t\t")
+    snippet.append_placeholder("any of them")
+    return snippet
+
+
+def _generate_string_snippet(snippet: SnippetString) -> SnippetString:
+    snippet.append_text("\tstrings:\n\t\t")
+    snippet.append_placeholder(r"\$name")
+    snippet.append_text(" = ")
+    snippet.append_choice(('"string"', "/regex/", "{ HEX }"))
```

Comment:
I always wanted to provide a tabstop in these choices so user's could easily swap out the `string`, `regex`, and `HEX` text, but I could never figure out a way to do it with the VSCode API. It's a small quality improvement, but that may be something worth looking into more